Sources/App/Auth/Authenticators/AdminAuthenticator.swift

```diff
@@ -0,0 +1,37 @@
+//
+//  File.swift
+//  
+//
+//  Created by Joe Williams on 16/11/2021.
+//
+
+import Foundation
+import Vapor
+import Fluent
+
+class AdminAuthenticatable: AsyncBearerAuthenticator {
+    
+    enum RoleFailure: Error {
+        case notAuthenticated
+        case notAdmin
+    }
+    
+    func authenticate(bearer: BearerAuthorization, for request: Request) async throws {
+        guard let token = try await UserToken.query(on: request.db).filter(\.$value == bearer.token).first() else {
+            throw Abort(.unauthorized)
+        }
+        
+        guard token.isValid else {
+            try await token.delete(force: true, on: request.db)
+            throw Abort(.unauthorized)
+        }
+        
+        let user = try await token.$user.get(on: request.db)
+        guard user.role == .admin else {
+            throw Abort(.unauthorized)
+        }
+        
+        request.auth.login(token)
```

Comment:
Do you need both of these? I thought you only needed the `.auth.login(user)` but could be entirely wrong 🤔

Reply:
So this is used to protect admin only routes. The rationale was the CMS would only be visible and accessible for users who were an admin, and you could get granular with that like a `SpeakerAuthenticatable` so speakers can access an area to upload talk resources etc